compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -3021,29 +3025,39 @@ impl<'a> Parser<'a> {
             }
             self.bump();
         }
+
         let mut err = self.dcx().struct_span_err(spans, "encountered diff marker");
-        err.span_label(start, "after this is the code before the merge");
-        if let Some(middle) = middlediff3 {
-            err.span_label(middle, "");
-        }
+        match middlediff3 {
+            // We're using diff3
+            Some(middlediff3) => {
+                err.span_label(
+                    start,
+                    "between this marker and `|||||||` is the code that we're merging into",
+                );
+                err.span_label(middlediff3, "between this marker and `=======` is the base code (what the two refs diverged from)");
+            }
+            None => {
+                err.span_label(
+                    start,
+                    "between this marker and `=======` is the code that we're merging into",
+                );
+            }
+        };
+
         if let Some(middle) = middle {
-            err.span_label(middle, "");
+            err.span_label(middle, "between this marker and `>>>>>>>` is the incoming code");
         }
         if let Some(end) = end {
-            err.span_label(end, "above this are the incoming code changes");
+            err.span_label(end, "this marker concludes the conflict region");
         }
-        err.help(
-            "if you're having merge conflicts after pulling new code, the top section is the code \
-             you already had and the bottom section is the remote code",
-        );
-        err.help(
-            "if you're in the middle of a rebase, the top section is the code being rebased onto \
-             and the bottom section is the code coming from the current commit being rebased",
-        );
+        err.help("conflict markers indicate that a merge was started but could not be completed due to merge conflicts");
+        err.help("to resolve a conflict, keep only the code you want and then delete the lines containing conflict markers");
```

Comment:
My only concern with the wording change is that the meaning change between a merge conflict when merging and a merge conflict when rebasing isn't being communicated.

Reply:
I'm ok with all the other changes. Let's do the following
```suggestion
        err.note(
            "conflict markers indicate that a merge was started but could not be completed due \
             to merge conflicts\n\
             to resolve a conflict, keep only the code you want and then delete the lines \
             containing conflict markers",
        );
        err.help(
            "if you're having merge conflicts after pulling new code, the top section is the code \
             you already had and the bottom section is the remote code\n \
             if you're in the middle of a rebase, the top section is the code being rebased onto \
             and the bottom section is the code coming from the current commit being rebased",
        );
```

It won't help with the concerns about verbosity, but I prefer to err on the side of giving too much information as opposed to too little.

Reply:
Are those `\n\` intentional? You could left them out and have the same effect but I assume that they're for clarity?

Reply:
The `\` also eliminates the leading whitespace in the next line, otherwise you end up with text like

```
you already had and the bottom section is the remote code
             if you're in the middle of a rebase, the top section is the code being rebased onto
```